tests/integration/bundle/test_bundle.py

```diff
@@ -0,0 +1,132 @@
+            deploy_postgres_k8s_bundle(ops_test),
+            ops_test.model.deploy("finos-waltz-k8s", application_name=FINOS_WALTZ, channel="edge"),
+        )
+        await ops_test.model.wait_for_idle(apps=[PG, PGB, FINOS_WALTZ], timeout=600)
+        await ops_test.model.add_relation(f"{PGB}:db", f"{FINOS_WALTZ}:db")
+        wait_for_relation_joined_between(ops_test, PGB, FINOS_WALTZ)
+        await ops_test.model.wait_for_idle(
+            apps=[PG, PGB, FINOS_WALTZ], status="active", timeout=600
+        )
+        backend_relation = get_backend_relation(ops_test)
+        pgb_user, pgb_password = await get_backend_user_pass(ops_test, backend_relation)
+        await check_database_creation(ops_test, "waltz", pgb_user, pgb_password)
+
+
+@pytest.mark.bundle
+async def test_discover_dbs(ops_test: OpsTest):
+    """Check that proxy discovers new members when scaling up postgres charm.
+
+    Since there's only one readonly endpoint, we can only really test this by querying whether the
+    endpoint exists or not depending on the number of postgres replicas.
+    """
+    pgb_unit = f"{PGB}/0"
+    backend_relation = get_backend_relation(ops_test)
+    backend_databag = await get_app_relation_databag(ops_test, pgb_unit, backend_relation.id)
+    assert backend_databag.get("read-only-endpoints") is None
+
+    await scale_application(ops_test, PGB, 3)
+    await ops_test.model.wait_for_idle()
+    await scale_application(ops_test, PG, 3)
+
+    updated_backend_databag = await get_app_relation_databag(
+        ops_test, pgb_unit, backend_relation.id
+    )
+    assert (
+        updated_backend_databag.get("read-only-endpoints") is not None
+    ), f"read-only-endpoints not populated in updated backend databag - {updated_backend_databag}"
+
+
+@pytest.mark.bundle
+async def test_kill_pg_primary(ops_test: OpsTest):
+    """Kill postgres primary, check that all proxy instances switched traffic for a new primary."""
+    # get connection info
+    finos_unit_name = f"{FINOS_WALTZ}/0"
+    finos_relation = get_connecting_relations(ops_test, PGB, FINOS_WALTZ)[0]
+    finos_databag = await get_app_relation_databag(ops_test, f"{FINOS_WALTZ}/0", finos_relation.id)
+    connstr = finos_databag.get("master")
+    assert connstr is not None, f"databag incorrectly populated, \n {finos_databag}"
+
+    pgb_unit = get_leader(ops_test, PGB).name
+    pgb_unit_address = await get_unit_address(ops_test, pgb_unit)
+    conn_dict = pgb.parse_kv_string_to_dict(connstr)
+    conn_dict["host"] = pgb_unit_address
+    connstr = pgb.parse_dict_to_kv_string(conn_dict)
+
+    # Get postgres primary through action. Note that postgres primary != postgres leader.
+    unit_name = ops_test.model.applications[PG].units[0].name
+    action = await ops_test.model.units.get(unit_name).run_action("get-primary")
+    action = await action.wait()
+    primary = action.results["primary"]
+    old_primary_address = await query_unit_address(connstr)
```

Comment:
Worth checking if action passed first. 

If this test fails, `KeyError` isn't informative. Maybe check status-code of action or something?

Reply:
Good point, I've verified `action.results.get("primary")` contains what we want, and I've logged the action results in the event it doesn't contain what we want.